pkgs/checks/README.md

```diff
@@ -1,23 +1,120 @@
+```
+
+Some expectations take arguments which are themselves expectations to apply to
+other values. These expectations take `Condition` arguments, which check
+expectations when they are applied to a `Subject`. The `ConditionSubject`
+utility acts as both a condition and a subject. Any expectations checked on the
+value as a subject will be recorded and replayed when it is applied as a
+condition. The `it()` utility returns a `ConditionSubject`.
+
+```dart
+checkThat(someList).any(it()..isGreaterThan(0));
+```
+
+Some complicated checks may be difficult to write with parenthesized awaited
+expressions, or impossible to write with cascade syntax. There are `which`
+utilities for both use cases which take a `Condition`.
+
+```dart
+checkThat(someString)
+  ..startsWith('a')
+  // A cascade would not be possible on `length`
+  ..length.which(it()
+    ..isGreatherThan(10)
+    ..isLessThan(100));
+
+await checkThat(someFuture)
+    .completes()
+    .which(it()..equals(expectedCompletion));
+```
+
+# Writing custom expectations
+
+Expectations are written as extension on `Subject` with specific generics. The
+library `package:checks/context.dart` gives access to a `context` getter on
+`Subject` which offers capabilities for defining expectations on the subject's
+value.
+
+The `Context` allows checking a expectation with `expect`, `expectAsync` and
+`expectUnawaited`, or extracting a derived value for performing other checks
+with `nest` and `nestAsync`. Failures are reported by returning a `Rejection`,
+or an `Extracted.rejection`, extensions should avoid throwing exceptions.
+
+Descriptions of the clause checked by an expectations are passed through a
+separate callback from the predicate which checks the value. Nesting calls are
+made with a label directly. When there are no failures the clause callbacks are
+not called. When a `Condition` is described, the clause callbacks are called,
+but the predicate callbacks are not called. Conditions can be checked against
+values without throwing an exception using `softCheck` or `softCheckAsync`.
+
+```dart
+extension CustomChecks on Subject<CustomType> {
+  void someExpectation() {
+    context.expect(() => ['meets this expectation'], (actual) {
+      if (_expectationIsMet(actual)) return null;
+      return Rejection(which: ['does not meet this expectation']);
+    });
+  }
+
+  Subject<Foo> get someDerivedValue =>
+      context.nest('has someDerivedValue', (actual) {
+        if (_cannotReadDerivedValue(actual)) {
+          return Extracted.rejection(which: ['cannot read someDerivedValue']);
+        }
+        return Extracted.value(_readDerivedValue(actual));
+      });
+
+  // for field reads that will not get rejected, use `has`
+  Subject<Bar> get someField => has((a) => a.someField, 'someField');
+}
+```
+
+# Trying Checks as a Preview
 
 1.  Replace the existing `package:test/test.dart` import with
     `package:test/scaffolding.dart`.
 
+1.  Add an import to `package:checks/checks.dart`.
```

Comment:
We should mention adding the dependency also probably?

Reply:
I was kind of treating that as implied now that it will be on pub. It was helpful before to help with using git constraints, but I don't think we should pub info about pub deps into pub package readmes generally.

Reply:
It is a required part of the process though, and while its implied for us I think it won't be for everybody?